Pipeline State Objects revolution

sectopod opened the pull request:
Requires https://github.com/gfx-rs/draw_state/pull/34
Fixes #815, fixes #842, fixes #823, fixes #761, fixes #722, fixes #703, fixes #694, fixes #463
Closes #817, closes #640, closes #630, closes #575, closes #571, closes #322, closes #121
Touches #698, #331, #311, #284, #161

Breaking changes:
- `Batch` concept is gone, PSO is the new paradigm
- `Mesh` is gone, but `Slice` is still with us, even if slightly modified
- shader parameters are superseded by the PSO configuration
- VAO, FBO, `Surface`, `Plane` are all gone, they are now hidden GL implementation details
- Users are required to operate on the strongly-typed target/resource views

TODO (note: this is a vastly incomplete list of what's actually done):
- [x] PSO resource - low level
  - [x] resource
  - [x] binding
  - [x] creation
- [x] High level
  - [x] `RenderTargetView` and `DepthStencilView`
    - [x] low-level bindings
    - [x] raw creation
    - [x] typed creation
    - [x] hook up to PSO
  - [x] `ShaderResourceView` and `UnorderedAccessView`
    - [x] low-level bindings
    - [x] raw creation
    - [x] typed creation
    - [x] hook up to PSO
  - [x] move core into a separate lib (#703)
  - [x] remove the Copy constraint from the resources
  - [x] layout matching
      - [x] vertex attributes
      - [x] shader data
      - [x] pixel targets
  - [x] strongly-typed initialization
  - [x] hook to the `Renderer` for drawing
- [x] helper macros
  - [x] structure definition
  - [x] PSO
  - [x] PSO with static init
  - [x] texture format
  - [x] macros tests
- [x] Universal data format spec
    - [x] serialized form
    - [x] strongly-typed form
    - [x] replace `tex::Format`
    - [x] replace `attrib::Format`
- [x] GL backend
    - [x] rewrite `init()` function for windowing backends
    - [x] rewrite `CommandBuffer` implementation to do the heavy lifting
    - [ ] ~~support for structured buffers~~ (can be postponed)
    - [x] Glutin window
    - [x] GLFW window
- [x] Switch the examples to use PSO
  - [x] blend
  - [x] cube
  - [x] deferred
  - [x] flowmap
  - [x] instancing
  - [x] mipmap
  - [x] performance
  - [x] shadow
  - [x] terrain
  - [x] triangle
  - [x] ubo_tilemap
- [x] Clean up
  - [x] remove old `Resources` subtypes
  - [ ] ~~remove shader programs~~
  - [x] remove old meshes and attribute formats
  - [x] remove old shader parameters
  - [x] remove old texture format
  - [x] remove batches
  - [x] remove VAO and FBO
  - [x] remove `Plane`, `Output`, and `Frame`
- [x] Post-clean
  - [x] rename `NewTexture` to `Texture`
  - [x] remove `primitive` from `mesh::Slice`
- [x] Documentation
  - [x] typed PSO
  - [x] universal format

Or, basically, rewrite the whole damn thing ;)

homu commented:
:umbrella: The latest upstream changes (presumably #829) made this pull request unmergeable. Please resolve the merge conflicts.

csherratt commented:
Sorry @sectopod for creating a conflicts.

sectopod commented:
@csherratt np!
Heads up - I'm going to change `ShaderParam::Resources` to be a generic parameter rather than an associated type. This would technically allow the same `ShaderParam` to be used with different devices in the same program.

sectopod commented:
I'm going through several iterations internally, and the progress with this PR is slow but steady. If you need to publish a new version for some other changes - please don't wait for me.
The way this PR is heading now changes quite a bit of perspective in these areas:
  1. Shader resources are serialized into data (`LinkMap`) and passed to the device for program linkage. This is the opposite from what we have now: program linked separately, `ProgramInfo` generated as a serialized form of shader demands, and then they are generically matched against user data.
  2. The concrete resource types, like vertex attributes, have their layout matching logic inside the hidden type (akin `VertexFormat`). This type has the names and formats of the data at compile time (static methods), and the actual linkage to the shader expectations at run time (index data). Vertex structure itself becomes purely data with no logic. This is quite different from the current design, where the generate `Link` structure is just data, and `Vertex` itself has all the logic.
  3. The PSO defines how the data is processed (inputs, outputs, algorithms, parameters, state), but doesn't have the data itself (unlike our old batches). There is going to be less run-time checks during rendering and stronger type guarantees. Prototype PSO definition:
```rust
pub struct PipelineState<
    R: d::Resources,
    I: VertexLayout<R>,
    P: ShaderParam<R>,
    E: PixelLayout<R>
>(d::handle::RawPipelineState<R>, d::PrimitiveType, I, P, E);
```

sectopod commented:
Why do we even separate vertex attributes from pixel targets from all the other shader data? If we follow #833, we can just define all the shader interface in a single struct like this one:
```rust
gfx_shader_link!( Shader {
    vertex: VertexBuffer<Vertex>, //regular vertex buffer, where `Vertex: VertexFormat`
    vertex_inst: VertexBuffer<Instance>, //instanced buffer, the instance rate is specified somewhere inside the `Instance` struct definition
    const_locals: ConstantBuffer<Local>, //constant buffer
    const_globals: ConstantBuffer<Global>, // another one
    tex_diffuse: TextureView<Dim2, Float4>, // 2D texture SRV
    tex_normal: TextureView<Dim2, Float3>, // same, but of `Float3` type
    sampler_linear: Sampler, // just a sampler, not sure if it would make sense to typify further here
    buf_noise: BufferView<Int4>, // structured buffer SRV
    buf_frequency: UnorderedView<Dim2, Int>, // 2D UAV buffer of `Int`
    pixel_color: TargetView<Dim2, Float4>, // output render target
});
```
Of course, each of the vertex buffers need to have a special struct behind them too (like the current `gfx_vertex`), but still it looks rather sleek to me.

kvark commented:
@sectopod great idea! so we are going to include attributes and targets into a struct what used to be `ShaderParam`. This would make all the `Output`, `Frame`, and even some `Mesh` stuff irrelevant and simplifies the user side of things: for a draw call, we'd only need a slice, a PSO, and this shader link block. It also simplifies strongly-typed PSO definition: it's only parametrized by the `Resources` and this `ShaderLink` type (maybe `ShaderData` would be better? or `PipelineData`?).

Overall, I think you are not just on the right track, but onto something big. Having all these `*View` types, strong PSO, unified shader link, and with all the old GL cruft removed - gfx would become a very different beast: better suited for next-gen APIs, safer, and friendlier to the users. Keep up!

sectopod commented:
It took me a fair bit of persistence to implement this scheme, and it could work if finished (got several large commits locally, hesitating to push yet). The reason I stopped is because I feel that this design is incomplete.

Take the blend state for example: providing it separately in the `DrawState` struct, one per color attachment, has no link to the actual target views we provide in this shader link structure. This is not just redundant, it's also a failure to allow using a specific blend mode on a specific target. Similar, even if less severe, is the situation with the depth/stencil - it's possible to provide a depth view without a depth test and the opposite, at run-time, which is sub-optimal.

A logical conclusion for me is to somehow provide the blending, depth state and offset, and stencil state as a part of this [shader link](https://github.com/gfx-rs/gfx/pull/828#issuecomment-156771805) structure. This would mean we no longer need to pass `DrawState` for PSO creation, but instead some sort of extended `Primitive` configuration that would include the topology and multi-sampling (I'd call it `Rasterizer`). However, there is one big problem here - blend/depth/stencil can not be a part of this structure, since it only has the data needed for run-time execution. The configuration for PSO creation is derived from the types themselves (in the current iteration of the design). 

Basically, this shader link has 3 different structures underneath:
 1. PSO initialization structure. It has the rasterizer configuration, blend states, depth/stencil, and the shader names of all the fields. It's currently partly raw-data-driven (`create_pipeline_state(..., &DrawState, ...)`), partly user-defined at compile time (shader symbols), and partly implicitly derived from the types of the declaration fields (e.g. `RenderTargetView<R, T>` derives the target format from `T`).
2. PSO meta-data. It has all the register indices for the used fields, constructed based on the shader reflection. The structure is generated from the field types and managed automatically (though we still need to specify the name because of the current Rust macro system limitation - inability to create new identifiers).
3. PSO run-time data. It has the buffer/view handles of the stuff we are binding on each draw. The structure is the one we are providing to the shader link macro.

So, going back to blend/depth/stencil - we are at the crossroads. Here are some extreme routes we can take, or maybe we can figure out a reasonable compromise in-between (if only anyone would be able to follow my essays...).

1. Extreme types - provide all the initialization data with types. This would mean compile-time fixed depth/blend/stencil states, which is a rather strict limitation. It would also rely heavily on our ability to support more generic versions of the shader links in the macro, which would at least allow some variability of these states. It would also involve a bigger gap between the types we pass at run-time and the types of the declaration, which means more boilerplate code by the user...

2. Extreme data - provide all the init data with data. It could be achieved by generating this PSO init struct by our macro and requiring the user to fill it up and pass in for PSO creation. This structure would contain the symbol names too, so it would provide the highest flexibility we can ever get without compromising the strong types safety.
The downside here is mainly the mental load on the user - now they've got to think about this init structure as well, which format they don't observe visually and can only guess (with compiler help) upon filling. For example, `RenderTargetView<R, T>` in the run-time declaration would imply something like `(&'a str, Option<Blend>, ColorMask)` it the init struct.

Any ideas/opinions on how to proceed? Am I even nearly close to an efficient design here? Without further instructions I'll go with extreme data path to see where the rabbit whole leads...

kvark commented:
@sectopod nice to see steady progress! It's exciting to see how far we can reach with this shader link concept in terms of safety and usability. BTW, I vote to rename it to `PipelineData` (and `gfx_pipeline_data` correspondingly).

As for your main question, I believe a reasonable compromise between data and types w.r.t `DataLink` would be to have as little types as needed to ensure no possible errors happening at run-time binding. Blend state, for example, is impossible to screw up at run time, so we can have it as data. Symbol names - too - should be data. SRV/DSV/RTV formats though have to match, so we should be having them typed.

kvark commented:
Ok, I've been thinking about it some more... I still don't know *how* to get this right, but at least I know how to definitely get this wrong ;)

First off, using types only for PSO initialization (`Extreme types`) is a no-go. It harms the flexibility too much (inability to tweak the blend state, depth state, or even the shader-exposed name).

Secondly, generating the PSO init struct in a macro and then expecting the user to instantiate it is another no-go. Everything that user is supposed to create should be either explicitly documented, or declared by the user themselves.

One possible research direction that is left after cutting these paths is - expect the user to provide both init and bind parts explicitly. Reminder: the "init" part is the one passed for PSO creation, the "bind" part is the one provided for each draw call.
```rust
color: gfx::RenderTargetView<R, Float4> | (&str, gfx::BlendState, gfx::ColorMask),
```
This will generate the `color` field in all init/meta/bind structs, where the user will need to provide the `RTV<Float4>` for binding, and `(name, blend, color_mask)` for initialization. If rust typesystem is flexible enough, we'll be able to enforce the `Float4` to be a blendable format too. In other words, we should use the init types to help define what the expectactions for the bind ones are: having a depth state means our DSV should have the depth component, having the blend state means our RTV should be blendable, etc.

kvark commented:
Another idea. If we don't take the shader variable names into account, the only data that we need to pass for the PSO creation is related to render targets: blend states, color masks, target formats, depth/stencil. This fact makes the targets somewhat different from the rest of our magical struct. Also the fact that there can be only one depth/stencil view, while there can be any repeating number of other types of entries. Perhaps, we should not include the render targets into this struct then? Treating them separately would make the rest of the struct clean enough and reduce the confusion. For the targets, we can use nominal types (tuples) to avoid user-specified names.

sectopod commented:
@kvark hmm, looks like there is no clearly superior solution in sight. Specifying both the bind data and init data types is too verbose to my taste (I'd rather have the some staff magically derived). Moving render target stuff out would complicate the interface... 
I don't know what to do yet, but I tried to systematize all the data that we need to support for PSO magical struct, minus all the rasterizer stuff. Hope this will help:

prototype 							| init data 				| meta	| bind data
----------------------------------------------------- | ---------------------------------- | ---------- | ----------------------------
`VertexBuffer<T: Structure>`			| `(T::init, InstanceRate)` 	| [v#]	| `BufferHandle<R, T>`
`ConstantBuffer<T: Structure>`			| `(T::init, &str)`			| c#		| `BufferHandle<R, T>`
`Constant<T: Uniform>`				| `(&str, T::format)`		| loc		| `T`
`Texture<T: TextureFormat>`			| `(&str, T::format)`		| t#		| `ShaderResourceView<R, T>`
`Sampler`							| `&str`					| s#	| `Sampler<R>`
`UnorderedBuffer<T: BufferFormat>`	| `(&str, T::format)`		| u#	| `UnorderedAccessView<R, T>`
`RenderTarget<T: RenderFormat>`	| `(&str?, T::format, ColorMask)`		| index	| `RenderTargetView<R, T>`
`BlendTarget<T: BlendFormat>`		| `(&str?, T::format, ColorMask, BlendState)`	| index	| `RenderTargetView<R, T>`
`DepthTarget<T: DepthFormat>`		| `(T::format, DepthState)`	| 	| `DepthStencilView<R, T>`
`StencilTarget<T: StencilFormat>`		| `(T::format, StencilState)`	| 	| `DepthStencilView<R, T>`
`DepthStencilTarget<T: DepthStencilFormat>`	| `(T::format, DepthState, StencilState)`	| 	| `DepthStencilView<R, T>`

sectopod commented:
I believe I solved the main problem with PSO internal type hierarchy and getting it all defined by helper macros. The biggest issue right now is generating the identifier names in those macros. Since `concat_idents!` is useless, I have to demand all the names provided by the user.

I tried a different approach - to create a new internal module with a user-given name and then just use some predefined simple names inside. However, this turns out to be playing very badly with user-given parameters, which the user expects to be visible, while jumping into a sub-module instantly changes the visibility scope...

As always, any ideas are welcome. It would be nice to get this right from the start, since it affects the API verbosity and complexity.

sectopod commented:
All the PSO support is complete, including the strongly-typed views/textures and the unified data format. It turned out I could re-use much less stuff from our previous implementation, so it more and more looks like a rewrite of the whole library. Of course, there is more stuff to polish (especially on the GL backend side), but I think we are ready to proceed to phase 2 now: porting the examples.

sectopod commented:
Here is another update. We are at the end of phase-2: all the examples are now ported, minus some stuff left to fix in the mipmapping example. Moreover, the GL backend got improved massively, with the new format specification supported natively there. So, in other words, no rough spots there.

You can see how the examples became sleeker. There is less magic wrappers and more direct control (e.g. a window is no longer held inside a stream). Some new ergonomic hotspots emerged though, which need to be thought of:
- `gfx::VertexBuffer` initialization is almost always used with `PER_VERTEX`. Optimally, we'd want a separate `gfx::instanceBuffer` with no init data for either. This seems relatively easy to address.
- `gfx_pipeline_init` is used everywhere, and the need for `Init::new()` call when PSO is created. Not sure if we can do much outside of renaming stuff.
- `gfx::RenderTarget` always needs a color mask. We could have a separate `MaskedTarget` for that, but this solution faces a bit of an implementation problem (of how to reduce code duplication in `render/pso.rs`).

The end of phase-2 means we can now start cleaning up all the old stuff (phase-3), at last! This should not only be fun, but also liberating - to improve the compile/link speed, which drags down my productivity at times (especially it used to, before I moved the core into a separate crate).

One thing needs to be reiterated on though, in the meantime - the types of SRV/UAV. Currently, they are typed by `gfx::format::Formatted`, which includes the surface type. With this implementation, a PSO declaration can't seamlessly accept both `SRV<U8Norm>` and `SRV<f32>`, for example, while this should be perfectly possible. One (quick) solution is to bring up the casting functions into public space for SRV/UAV, so that the user can just cast upon assigning to pso data. Another (better) solution would be to re-think the new format specification and detach the view format from the surface format.

bjz commented on examples/cube/main.rs:
```diff
@@ -18,39 +18,45 @@ extern crate gfx;
-gfx_parameters!( Params {
-    u_Transform@ transform: [[f32; 4]; 4],
-    t_Color@ color: gfx::shade::TextureParam<R>,
+gfx_pipeline_init!( PipeData PipeMeta PipeInit {
```
It's a shame about this long list of idents... I guess there's no way around it though... :(

sectopod replied:
Indeed. I've been trying to get everyones attention to this issue, but I suppose it requires a bit too much mental investment to even understand how it works. As it stands now, all 3 new types are (or can be) used directly by the user:
- `PipeData` is always needed to be created since it's required for every call
- `PipeMeta` is needed if one needs to store the PSO somewhere, the full type of which is `gfx::PipelineState<R, PipeMeta>`. Out of my head, `ubo_tilemap` example stores it.
- `PipeInit` is needed for PSO creation.

Perhaps, some of them can be re-routed via others by increasing the internal complexity of the type structure a little bit.

bjz replied:
Would be kind of nice to have anonymous structs for associated types here:

```rust
impl<'a> PipelineInit for Foo<'a> {
    type Meta = struct {
        ...
    };

    type Data = struct {
        ...
    };
}
```

Perhaps you could use modules here:

```rust
gfx_pipeline_init! {
    mod pipeline {
        ...
    }
}

// expands to:

pub mod pipeline {
    pub struct Init {
        ...
    }

    pub struct Meta {
        ...
    }

    pub struct Data {
        ...
    }
}
```

sectopod replied:
Thanks for trying to tackle this problem! I iterated on the separate module idea earlier:
> I tried a different approach - to create a new internal module with a user-given name and then just use some predefined simple names inside. However, this turns out to be playing very badly with user-given parameters, which the user expects to be visible, while jumping into a sub-module instantly changes the visibility scope...

However, I didn't consider/try doing `use super::*;` at the top of the new module. Perhaps, this would work? As a side effect it may require dependent types (vertex structures) to be declared as public or re-exported as public.

sectopod replied:
@bjz this worked after all. There are some hoops though, like the need to `pub use` instead of just `use` for the stuff used inside the macro, but overall I consider it a success. Please have a look at the latest code in this PR.

bjz commented on examples/performance/main.rs:
```diff
@@ -119,21 +136,16 @@ fn gfx_main(mut glfw: glfw::Glfw,
-                stream.draw(&batch.with(&slice, &data, &state))
-                      .unwrap();
+                data.transform = proj.mul_m(&view.mat)
+                                     .mul_m(&model).into_fixed();
```
Still on an old version of cgmath? :(

sectopod replied:
I bet this PR started before your grand cgmath changes ;)

bjz replied:
Yup! :)

sectopod commented:
I believe all the ergonomics concerns (at least the one mentioned in this thread) are now addressed. Overall, the code is getting polished nicely, and only a few bits left to do:
- better SRV/UAV typing
- documentation
- somebody needs to actually review this code

sectopod commented:
Hmm. I implemented better typing for SRV/UAV that allow us to avoid the casts (e.g. in deferred example), but now Rust forces me to explicitly specify the format type for the view creation functions in `Factory`, which is a little annoying. Looking into ways to simplify it to avoid explicit type annotations...

sectopod commented:
We are green, and my checklist is completed. All systems ready to go!

cmr commented:
@sectopod awesome work, been really excited about this new model! Do you have any recommendations for going about reviewing this code? 138 commits is a bit long, and I imagine at least some of it obsoleted by later commits. Maybe some strategic squashing?

homu commented:
:umbrella: The latest upstream changes (presumably #845) made this pull request unmergeable. Please resolve the merge conflicts.

sectopod commented:
@cmr thank you for the intent of reviewing the code! I understand it may take a while to scan through all the changes, and I don't think it's worth it. The systems I implemented in this PR went through a couple iterations of evolution, at least. It was one hell of an R&D work. Some of this history is reflected in the actual commits, some is not.

Thus, I'd suggest just looking at the final revision with this in mind:
  1. gfx core has been moved into a separate crate (and now I wonder how Travis succeeded, given this crate is not published?..)
  2. lots of old stuff is removed
  3. `Renderer` is renamed to `Encoder`, module name is also changed
  4. new `Texture` is not an evolution of the old one. It was developed separately and then replaced the old name.
  5. quite some bits of logic have migrated from gfx itself into the GL backend: command definition and `CommandBuffer` implementation, Surface concept, and others.

cmr commented:
Thanks! I'll spend time tomorrow (ie, after 9 hours from now) reviewing this.

bvssvni commented:
@sectopod Travis CI probably tested the paths in Cargo.toml, which does not require publishing the crates when building the local project.

kvark commented:
@sectopod I'm glad you managed to finish this epic task. It started as a little feature, and then turned into a rewrite of most of our old code, and re-design of our concepts from the ground up. At some point, I thought the feature creep would just eat you alive ;)

To recap, we drop all the "artificial" concepts we had (batches, frames, planes, meshes, etc), and replace them with a single yet powerful concept of a `PipelineState` (PSO). It's natural, since it closely follows API specification (DX12 + Metal). The `Texture` is just a pixel storage now, and we introduce the accompanying views (shader resource (SRV), unordered access (UAV)) to hook up to PSO. We also get the universal format specification for vertex buffers and texture surfaces.

Everything exists in two worlds: type-level, checked at compile time, and data-level, checked at run-time. Users are supposed to work on the type-level (strongly-typed PSO, views, formats), which gets automatically transcribed into the data-level to be consumed by the non-generic interfaces of `gfx_core` crate, which are implemented by the API backends.

I've been following the development closely and welcome the changes, very excited to see what we can do with the new architecture! I advice everyone interested in the project to have a look to get more familiar with the implementation. Perhaps, you will have some notes for the review too, in order to back up me and @cmr.

Trebgarta commented:
It is very nice to see that this is close to being merged. I am interested in this project so I want to ask this: When can I expect this to be merged, and should I expect more breaking changes in the future?

cmr commented:
I'm going to leave reviews in this comment, and append to it as I find more stuff to comment on.

------
https://github.com/sectopod/gfx-rs/blob/68f1265f52186aec73202c98182c4d24e5f85b54/src/core/src/factory.rs#L37

I'm a bit worried about overflow panics here, although maybe I shouldn't be. Using .wrapping_add and .wrapping_mul might be nice, if this function is always used with "safe" slices, to keep code size down. I'm also worried that it is marked `pub` but not `unsafe`, as in general this is very unsafe - it's equivalent to a transmute! - but is safe on "POD" types.
-------

kvark commented:
@Trebgarta I guess we are waiting for @cmr to finish the review, and for @sectopod to address the issues that @cmr finds.

> should I expect more breaking changes in the future

It would be a miracle if this single PR was final and stable, especially consider the fundamental property of the work. I want to publish 0.9 once this PR is merged, and then we'll start fixing all the rough corners with (hopefully) more people trying out the new interfaces. Then 0.10 should be published with all the fixes, and it should be stable.

sectopod commented:
@cmr please enumerate the points outlined so that I can refer to them.
About `cast_slice` - I forced `Copy` bound on the arguments, which should make it truly safe now.
I don't understand the point about overflow and `wrapping_*` usage there. Why would `mem::size_of` overflow ever?

kvark commented:
@niconii suggested to use the following pattern for our PSO struct macro:
```rust
    ($name:ident [ $($t:tt)* ] {
        $( $field:ident : $ty:ty ,)*
    }) => {
        as_item!{struct $name < $($t)* > {
            $( $field: $ty ,)*
        }}
    };
```
Main tricks are:
  1. Using square brackets (`[` and `]`)
  2. Using bare tokens (`tt`)
  3. Adding `as_item!` on top

It would allow us to have generic parallelization of PSO, which is crucial to get some flexibility for the user, given how strict the current type system bounds are.

sectopod commented:
@kvark  sadly this doesn't seem to be enough. We need to not only provide the generic bounds in a macro but also refer to complete types. I.e. in addition to `struct A<Q:Qu,W:Wu>` we need to have `A<Q,W>` somewhere, and having a token list doesn't help...

@niconii please see [this macro](https://github.com/sectopod/gfx-rs/blob/e2d1e3fbc0ebbfe39ee62c824deba79dace04e4d/src/render/src/macros/pso.rs#L29) if you are willing to help us further.

kvark commented:
@cmr are you still with us?

cmr commented:
@sectopod it shouldn't overflow, that's the point! If you don't use wrapping_add/wrapping_mul, the compiler will generate overflow checking code.

cmr commented:
Sorry if my review is a bit wide-ranging at this point, and not specific to your changes, it's devolved into a full code review.

cmr commented:
Left some new comments. After they are fixed, I want to do one last look-over and then :shipit:

sectopod commented:
@cmr thanks! you found quite a few important bits to fix.
(1), (2), (3), (4) - fixed
(6) doesn't work for me. Doing `cargo fmt` produces: `No such file or directory (os error 2)` followed by the usage info (for the `cargo fmt`). We can do it later on.
(7), (8), (9) - fixed
(5), (10) - left for TODO outside of this PR
(11) - doesn't matter for now, since we are only using VAOs whenever GL forces us to do that. EXT versions don't force it. Maybe if we find a better use for them, we'll need to query for EXT stuff too.
(12) not fixed intentionally. It's not a performance critical code, and using `mem_unitialized!` would only confuse us during any debugging.

cmr commented:
Who gets the honor of pushing the big green button? :)

sectopod commented:
homu!

cmr commented:
Oh right, I forget we use homu now!

sectopod commented:
I wonder if we should volunterely strip ourselves of the power of the green button. Just looking at it makes my skin goose. Only @homu should guard the star gate to upstream.

cmr commented:
@homu r+ 736b0c3

homu commented:
:hourglass: Testing commit 736b0c3 with merge ad9c403...

sectopod commented:
@cmr hohoho, it's happening!
btw, why did you select the commit explicitly? It's not like I can sneak anything in between your comment and @homu response

cmr commented:
habit, nothing personal ;)

homu commented:
:broken_heart: Test failed - [status](https://travis-ci.org/gfx-rs/gfx/builds/104018359)

sectopod commented:
@cmr here comes the problem: if I push a fix for beta/nightly and politely ask homu to retry, will it retry the same commit you specified?..

cmr commented:
Not actually sure...

sectopod commented:
Well, it should be fixed now. The error was real, and I'm glad that beta/nightly caught it.

cmr commented:
@homu r+ 3f92272

homu commented:
:hourglass: Testing commit 3f92272 with merge 64d8fce...

homu commented:
:sunny: Test successful - [status](https://travis-ci.org/gfx-rs/gfx/builds/104020192)

cmr commented:
:confetti_ball: :tada: Great work @sectopod

gaudecker commented:
Oh wow, this finally merged? Maybe I should try to raise Rust Theft Auto back from the dead and try this out :party:

kvark commented:
Hooray! :fireworks: 
@gaudecker this should be fun ;) Don't forget to check up with us on gitter in case you have any questions.
